### PR TITLE
feat: added `HoleCard` and `CommunityCard` types

### PR DESCRIPTION
**Pull request by dev-laky**
This PR introduces new card types/classes for better type control and oop pattern.
In addition, this PR covers new tests in `GameTest` and `PlayerTest` for higher line and mutation coverage.
Lastly, `HandRank` enum class was improved by adding numeric values for each rank for better comparability. 

fyi @timwiddo :
- Your new branch for updating the `HandEvaluator` object will be synced with the content of this PR in order to provide the new types/classes
- I will open a branch and sync it for you ;)

#### Latest Updates - 9th May 🚨
As @juliankuecken adressed a few issues concerning the blind and related bet logic, I decided to implement `raiseBetTo` and `call` function in order to provide these function to work with initializing the blinds inside `Game`.

I decided to work with `raiseBetTo` for the raising of the bets.
The name states clear that this function and logic behind it is always raising TO an specific amount, not BY an amount. 
With this logic it's easier for us to avoid any calculation mistakes by dealing with sums/diffs.

#### What didn't fit into the PR?
We need to extend the bet logic by `check`, `allIn`, `fold`. 
This is going to take place under the following Issue:
- #35

We also need to adjust the test coverage for the bet logic, which will be a included task in #35

### CODE REVIEW

**Review by timwiddo — APPROVED** [review]
Wonderful! Will update the evaluator asap.

**Comment by dev-laky** [comment]
> lgtm, but for better looking tests we could use the @beforeeach setup() for the tests so you dont have to initialize a game or player in every test. Furthermore we could set the blinds in the Game class since this is also true with real Poker tables.

@scoutah you are right, I will have a look on the `@beforeeach` annotation for testing to reduce the repetitive usage of the `MockGame`.

As we only need a mocked game in these tests, there is no need to provide any blinds.
The game logic is totally insignificant here because we only need a representative mock class. :)

File: src/test/kotlin/hwr/oop/projects/peakpoker/core/game/GameTest.kt [hunk 1]
```diff
@@ -96,6 +96,203 @@ class GameTest : AnnotationSpec() {
...
+
+        player1.raiseBet(30)
+        player2.raiseBet(30)
+        player3.raiseBet(30)
+
+        assertThat(testGame.getHighestBet()).isEqualTo(30)
+    }
+
+    @Test
+    fun `getHighestBet works with zero bets`() {
```
Comment (juliankuecken) [inline src/test/kotlin/hwr/oop/projects/peakpoker/core/game/GameTest.kt:184]: Should `getHighestBet()` not return 20 here? This would be the only valid consequence since we agreed upon smallBlind and bigBlind being automatically deducted.

File: src/test/kotlin/hwr/oop/projects/peakpoker/core/game/GameTest.kt [hunk 2]
```diff
@@ -96,6 +96,203 @@ class GameTest : AnnotationSpec() {
...
+    @Test
+    fun `getHighestBet handles edge cases with equal bets`() {
+        val player1 = Player("Hans")
+        val player2 = Player("Peter")
+        val player3 = Player("Max")
+        val testGame = Game(1040, 10, 20, listOf(player1, player2, player3))
+
+        player1.raiseBet(30)
+        player2.raiseBet(30)
+        player3.raiseBet(30)
```
Comment (juliankuecken) [inline src/test/kotlin/hwr/oop/projects/peakpoker/core/game/GameTest.kt:178]: ## Problem
If we say that each player raises TO his final bet (so if my current bet is 100 and i say raiseBet(200), it raises TO 200 instead of 100+200=300), lines [133-135](https://github.com/dev-laky/PeakPoker/pull/31/files#diff-02258a5bfcf7e4858f63a25c1bf692f3bc9324a0ce8b6743bec2ecefe8df1052R162-R168) are problematic.

## Reason
In these lines, the third player is be able to raise his bet to 20, even though the player before him has betted 30 (which is a fatal break of the poker rules).

File: src/test/kotlin/hwr/oop/projects/peakpoker/core/game/GameTest.kt [hunk 3]
```diff
@@ -96,6 +96,203 @@ class GameTest : AnnotationSpec() {
...
+
+        assertThat(testGame.getHighestBet()).isEqualTo(30)
```
Comment (juliankuecken) [inline src/test/kotlin/hwr/oop/projects/peakpoker/core/game/GameTest.kt:180]: ## Quickest fix
The quickest fix to the problem mentioned above is the following:
```suggestion
        assertThat(testGame.getHighestBet()).isEqualTo(90)
```
This is of course since (30+30+30=90).

File: src/test/kotlin/hwr/oop/projects/peakpoker/core/game/GameTest.kt [hunk 4]
```diff
@@ -96,6 +96,203 @@ class GameTest : AnnotationSpec() {
...
+        val player1 = Player("Hans")
+        val player2 = Player("Peter")
+        val player3 = Player("Max")
+        val testGame = Game(1042, 10, 20, listOf(player1, player2, player3))
+
+        player1.raiseBet(10)
+        player2.raiseBet(30)
+        player3.raiseBet(10)
+
+        assertThat(testGame.getHighestBet()).isEqualTo(30)
```
Comment (juliankuecken) [inline src/test/kotlin/hwr/oop/projects/peakpoker/core/game/GameTest.kt:204]: Same as above:
```suggestion
        assertThat(testGame.getHighestBet()).isEqualTo(50)
```
(10+30+10=50)

File: src/test/kotlin/hwr/oop/projects/peakpoker/core/game/GameTest.kt [hunk 5]
```diff
@@ -96,6 +96,203 @@ class GameTest : AnnotationSpec() {
...
+    fun `getPot returns correct value`() {
+        val testGame = Game(
+            1061, 10, 20,
+            listOf(Player("Hans"), Player("Peter"), Player("Max"))
+        )
+
+        testGame.pot = 100
+
+        assertThat(testGame.pot).isEqualTo(100)
+    }
```
Comment (juliankuecken) [inline src/test/kotlin/hwr/oop/projects/peakpoker/core/game/GameTest.kt:238]: No, this is fatal: This is not the way the pot should be returned and i imagined the functionality. You must call the `calculatePot()` function on the game. So:
```suggestion
    @Test
    fun `getPot returns correct value`() {
        val testGame = Game(
            1061, 10, 20,
            listOf(Player("Hans"), Player("Peter"), Player("Max"))
        )

        testGame.calculatePot() = 100

        assertThat(testGame.pot).isEqualTo(100)
    }
```

**Review by juliankuecken — APPROVED** [review]
Read through my comments, discuss the rest in lesson!

File: src/test/kotlin/hwr/oop/projects/peakpoker/core/cards/CommunityCardsTest.kt [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+            Card(Suit.SPADES, Rank.ACE),
+            Card(Suit.CLUBS, Rank.KING),
+            Card(Suit.HEARTS, Rank.QUEEN),
+            Card(Suit.HEARTS, Rank.JACK)
+        )
+
+        assertThatThrownBy { CommunityCards(cards, mockGame) }
+            .isInstanceOf(IllegalArgumentException::class.java)
+            .hasMessageContaining("exactly five cards")
+    }
```
[the remark this answers is not in the document]
Comment (dev-laky) [reply src/test/kotlin/hwr/oop/projects/peakpoker/core/cards/CommunityCardsTest.kt:69]: it does... 🫠

**Review by scoutah — APPROVED** [review]
lgtm

**Review by timwiddo — APPROVED** [review]
LGTM, further Questions asked on Whatsapp!

**Review by juliankuecken — CHANGES_REQUESTED** [review]
## General
Very good improvements since my last review, thank you for that. For further notes, look into my review comments.

## Improvements
There are some test-scenarios missing, that came to my mind during review. For that, see issue #38 .

File: src/main/kotlin/hwr/oop/projects/peakpoker/core/game/Game.kt [hunk 7]
```diff
@@ -14,13 +14,23 @@ class Game(
...
     var pot: Int = 0
     val communityCards: CommunityCards = CommunityCards(emptyList(), this)
-    var currentPlayerIndex: Int = 2
+
+    // Will be = 2 after "blind" init
+    var currentPlayerIndex: Int = 0
+
+    fun makeTurn() {
+        currentPlayerIndex = (currentPlayerIndex + 1) % playersOnTable.size
+    }
```
[the remark this answers is not in the document]
Comment (dev-laky) [reply src/main/kotlin/hwr/oop/projects/peakpoker/core/game/Game.kt]: Solved by commit [1d56698](https://github.com/dev-laky/PeakPoker/pull/31/commits/1d5669890bc1ebe45393623178103b1577ac3fac)

File: src/main/kotlin/hwr/oop/projects/peakpoker/core/player/Player.kt [hunk 8]
```diff
@@ -29,16 +30,53 @@ class Player(
...
+            """
+            Consider if check is needed inside of Player?
+            What could be the use case for this?
+            Otherwise remove, handle necessary actions in Game. 
+            
+            Quick thought:
+            Since nothing needs to be changed of bet or chips, 
+            all other properties of Player can be accessed through Game.
+            So right know I don't think we need this here.
+            """
```
[the remark this answers is not in the document]
Comment (dev-laky) [reply src/main/kotlin/hwr/oop/projects/peakpoker/core/player/Player.kt:71]: yes

File: src/main/kotlin/hwr/oop/projects/peakpoker/core/game/Game.kt [hunk 9]
```diff
@@ -42,11 +52,70 @@ class Game(
...
+     *
+     * @param player The player who is raising their bet
+     * @param chips The total amount to bet (not the additional amount)
+     * @throws IllegalStateException If the bet is not higher than the current highest bet,
+     *                               or if it's not the player's turn
+     */
+    fun raiseBetTo(player: Player, chips: Int) {
+        when {
+            chips < 0 -> throw IllegalArgumentException("Bet amount must be positive")
+            chips <= getHighestBet() -> throw IllegalStateException("Bet must be higher than the current highest bet")
```
[the remark this answers is not in the document]
Comment (dev-laky) [reply src/main/kotlin/hwr/oop/projects/peakpoker/core/game/Game.kt:73]: We just use the function name. No need for `this` since we are already inside the class scope.

File: src/main/kotlin/hwr/oop/projects/peakpoker/core/player/Player.kt [hunk 10]
```diff
@@ -29,16 +30,53 @@ class Player(
     }
 
     fun assignHand(cards: HoleCards) {
+        require(cards.cards.size == 2) { "A player must have exactly 2 hole cards" }
```
[the remark this answers is not in the document]
Comment (dev-laky) [reply src/main/kotlin/hwr/oop/projects/peakpoker/core/player/Player.kt:33]: Will be done by #38

File: src/main/kotlin/hwr/oop/projects/peakpoker/core/player/Player.kt [hunk 11]
```diff
@@ -29,16 +30,53 @@ class Player(
...
+     * between the new bet and current bet from the player's chips.
+     *
+     * @param chips The total amount to bet (not the additional amount)
+     */
+    fun raiseBetTo(chips: Int) {
+        require(chips >= 0) { "Bet amount must be positive" }
+
+        this@Player.chips -= chips - bet
+        bet = chips
+    }
```
[the remark this answers is not in the document]
Comment (dev-laky) [reply src/main/kotlin/hwr/oop/projects/peakpoker/core/player/Player.kt]: Fixed by [4e36efa](https://github.com/dev-laky/PeakPoker/pull/31/commits/4e36efa03eda05dac028084cb5161b6cd346309c)

**Review by juliankuecken — APPROVED** [review]
Looks good to me, thanks for the changes again. Enjoy your day! ;~)
